src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs

```diff
@@ -47,12 +47,10 @@ internal NewArrayInstruction(Type elementType)
         public override int Run(InterpretedFrame frame)
         {
             int length = ConvertHelper.ToInt32NoNull(frame.Pop());
-            if (length < 0)
-            {
-                // to make behavior aligned with array creation emitted by C# compiler
-                throw new OverflowException();
-            }
-            frame.Push(Array.CreateInstance(_elementType, length));
+            // To make behavior aligned with array creation emitted by C# compiler if length is less than
+            // zero we try to use it to create an array, which will throw an OverflowException with the
+            // correct localized error message.
+            frame.Push(length < 0 ? new int[length] : Array.CreateInstance(_elementType, length));
```

Comment:
In having this simulate what happens with compiled (whether from C# or a CIL-compiled lamdba) as closely as possible, if `newarr` is changed so the message contains the bad length this will get that change for free.